src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -0,0 +1,337 @@
+        time_between_requests.push_back(diff_ns.count());
+      }
+      return time_between_requests;
+    }
+
+    double CalculateAverage(std::vector<int64_t> values) {
+      double avg = std::accumulate(values.begin(), values.end(), 0.0) / values.size();
+      return avg;
+    }
+
+    double CalculateVariance(std::vector<int64_t> values, double average) {
+      double tmp = 0;
+      for (auto value : values) {
+        tmp += (value - average) * (value - average) / values.size();
+      }
+      double variance = sqrt(tmp);
+      return variance;
+    }
+
+    std::unique_ptr<LoadManager> CreateManager(PerfAnalyzerParameters params, bool is_sequence_model=false) {
+
+      std::unique_ptr<LoadManager> manager;
+      std::shared_ptr<cb::ClientBackendFactory> factory = std::make_shared<cb::MockClientBackendFactory>(stats_);
+      std::shared_ptr<ModelParser> parser = std::make_shared<MockModelParser>(is_sequence_model);
+
+      RequestRateManager::Create(
+        params.async, params.streaming, params.measurement_window_ms,
+        params.request_distribution, params.batch_size,
+        params.max_threads, params.num_of_sequences,
+        params.sequence_length, params.string_length,
+        params.string_data, params.zero_input, params.user_data,
+        params.shared_memory_type, params.output_shm_size,
+        params.start_sequence_id, params.sequence_id_range, parser,
+        factory, &manager);
+
+      return manager;
+    }        
+
+};
+
+
+/// Check that the correct inference function calls
+/// are used given different param values for async and stream
+///
+TEST_CASE("request_rate_infer_type") {
+  TestRequestRateManager trrm{};
+  bool async;
+  bool stream;
+
+  SUBCASE("async_stream") {
+    async=true;
+    stream=true;
+  }
+  SUBCASE("async_no_stream") {
+    async=true;
+    stream=false;
+  }
+  SUBCASE("no_async_stream") {
+    async=false;
+    stream=true;
+  }
+  SUBCASE("no_async_no_stream") {
+    async=false;
+    stream=false;
+  }
+
+  trrm.TestInferType(async, stream);
+}
+
+/// Check that the request distribution is correct for
+/// different Distribution types
+/// 
+TEST_CASE("request_rate_distribution") {
+  TestRequestRateManager trrm{};
+  PerfAnalyzerParameters params;
+
+  SUBCASE("constant") {
+    params.request_distribution = CONSTANT;
+  }
+  SUBCASE("poisson") {
+    params.request_distribution = POISSON;
+  }
+  trrm.TestDistribution(params);
+}
+
+/// Check that the request distribution is correct
+/// for the case where the measurement window is tiny
+/// and thus the code will loop through the schedule
+/// many times
+///
+TEST_CASE("request_rate_tiny_window") {
+  TestRequestRateManager trrm{};
+  PerfAnalyzerParameters params;
+  params.request_distribution = CONSTANT;
+  params.measurement_window_ms = 10;
+
+  SUBCASE("one_thread") {
+    params.max_threads = 1;
+  }
+  SUBCASE("odd_threads") {
+    params.max_threads = 9;
+  }
+  trrm.TestDistribution(params);
```

Comment:
Should TestDistribution take a parameter for how long it waits? It seems like the 10 ms window above in this test case has to be aware of what the sleep time is in TestDistribution in order for the test to work properly.

Reply:
added parameters